src/coreclr/jit/llvm.cpp

```diff
@@ -864,31 +867,82 @@ Value* Llvm::consumeValue(GenTree* node, Type* targetLlvmType)
     return finalValue;
 }
 
+Function* Llvm::getOrCreateLlvmFunction(const char* symbolName, GenTreeCall* call)
+{
+    Function* llvmFunc = _module->getFunction(symbolName);
+
+    if (llvmFunc == nullptr)
+    {
+        // assume ExternalLinkage, if the function is defined in the clrjit module, then it is replaced and an
+        // extern added to the Ilc module
+        llvmFunc =
+            Function::Create(getFunctionTypeForCall(call), Function::ExternalLinkage, 0U, symbolName, _module);
+    }
+    return llvmFunc;
+}
+
 llvm::Value* Llvm::buildUserFuncCall(GenTreeCall* call)
 {
-    const char* symbolName = (*_getMangledSymbolName)(_thisPtr, call->gtEntryPoint.handle);
+    llvm::FunctionCallee llvmFuncCallee;
 
-    (*_addCodeReloc)(_thisPtr, call->gtEntryPoint.handle);
-    Function* llvmFunc = _module->getFunction(symbolName);
+    if (call->gtCallType == CT_USER_FUNC)
+    {
+        const char* symbolName = (*_getMangledSymbolName)(_thisPtr, call->gtEntryPoint.handle);
 
-    std::vector<llvm::Value*> argVec = std::vector<llvm::Value*>();
+        (*_addCodeReloc)(_thisPtr, call->gtEntryPoint.handle);
+        Function* llvmFunc = getOrCreateLlvmFunction(symbolName, call);
+        llvmFuncCallee = llvmFunc;
+    }
+    else if (call->gtCallType == CT_INDIRECT)
+    {
+        Value* funcValue = getGenTreeValue(call->gtCallAddr);
+        if (call->gtEntryPoint.accessType == IAT_VALUE)
+        {
+            FunctionType* functionType = getFunctionTypeForCall(call);
 
-    for (GenTreeCall::Use& use : call->Args())
+            //TODO-LLVM: how best to detect there is no runtime lookup required when the target is a constant and not abstract or virtual
+            if (llvm::ConstantInt* llvmConstantInt = llvm::dyn_cast<llvm::ConstantInt>(funcValue)) {
+                void* methodHandle = (void*)(llvmConstantInt->getZExtValue());
+
+                const char* methodName = _getMangledSymbolNameFromHelperTarget(_thisPtr, methodHandle);
+                if (methodName == nullptr) // TODO-LLVM: abstract or virtual call
+                {
+                    failFunctionCompilation();
+                }
+
+                (*_addCodeReloc)(_thisPtr, methodHandle);
+
+                llvmFuncCallee = { functionType, getOrCreateLlvmFunction(methodName, call) };
```

Comment:
We are effectively doing "devirtualization for indirect calls" here, manually.

Why?

Is the issue that we need to call `(*_addCodeReloc)(_thisPtr, methodHandle);`? It does not seem right to have it be "pulled" from under the call, it should instead be registered at the logical `ldftn` point, when building the the icon node.

Reply:
Thanks, it's good that  you've questioned this, as it gives something concrete to talk about.  I didn't understand how to do the virtual calls so restricted myself to those calls that could be determined at runtime.  The `addCodeReloc` is there to make sure we get the target method compiled.  Happy to expand the PR to cover virtual calls but I didn't see how to make the slot lookup work, and I suspected I was missing something.  How do the other Ilc backends do this I wonder?

Reply:
@Jkotas @MichalStrehovsky  Maybe I can pull one of you in here please ?

Reply:
@yowl let me test my understanding here. Is the following true?

> `addCodeReloc` is only required for helpers. User methods are fine as they have been added as dependencies by the ILC's front-end.

Reply:
Ummm.... If I delete the call to `_addCodeReloc` , then it works.  Oh dear, sorry about that.  I still could use some help on the slot lookup side though.  Let me get some IR for that.

Reply:
Lets take https://github.com/dotnet/runtimelab/blob/dd8c43a825c41ccfc21a8ceefa79e232adbe02a1/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Debug.cs#L137-L138
The IR for this is
```cs
------------ BB01 [???..???), preds={} succs={BB02}

------------ BB02 [000..00F) (return), preds={BB01} succs={}
               [000017] ------------                 IL_OFFSET void   IL offset: 0x0
N001 (  1,  4) [000006] H-----------         t6 =    CNS_INT(h) int    0x420058 ftn
                                                  /--*  t6     int
N003 (  5,  7) [000008] DA----------              *  STORE_LCL_VAR int    V02 tmp1         d:1
N001 ( 14,  2) [000001] H-CXG-------         t1 =    CALL help byref  HELPER.CORINFO_HELP_READYTORUN_STATIC_BASE
N002 (  1,  1) [000002] ------------         t2 =    CNS_INT   int    4 Fseq[s_provider]
                                                  /--*  t1     byref
                                                  +--*  t2     int
N003 ( 15,  3) [000003] ---XG--N----         t3 = *  ADD       byref
                                                  /--*  t3     byref
N004 ( 17,  5) [000004] V--XGO-N----         t4 = *  IND       ref
               [000020] ------------        t20 =    CNS_INT   int    4
               [000021] ------------        t21 =    LCL_VAR   int    V04 tmp3         u:1
                                                  /--*  t21    int
                                                  +--*  t20    int
               [000022] ------------        t22 = *  ADD       int
                                                  /--*  t22    int
                                                  +--*  t4     ref
N006 ( 21,  8) [000014] -A-XGO----L-              *  STOREIND  ref
               [000023] ------------        t23 =    CNS_INT   int    4
               [000024] ------------        t24 =    LCL_VAR   int    V04 tmp3         u:1
                                                  /--*  t24    int
                                                  +--*  t23    int
               [000025] ------------        t25 = *  ADD       int
                                                  /--*  t25    int
N008 (  3,  2) [000015] ------------        t15 = *  IND       ref
               [000026] ------------        t26 =    CNS_INT   int    0
               [000027] ------------        t27 =    LCL_VAR   int    V04 tmp3         u:1
                                                  /--*  t27    int
                                                  +--*  t26    int
               [000028] ------------        t28 = *  ADD       int
                                                  /--*  t28    int
N009 (  3,  2) [000005] ------------         t5 = *  IND       ref
N010 (  3,  2) [000009] ------------         t9 =    LCL_VAR   int    V02 tmp1         u:1 (last use)
               [000029] ------------        t29 =    LCL_VAR   int    V04 tmp3         u:1
               [000030] ------------        t30 =    CNS_INT   int    8
                                                  /--*  t29    int
                                                  +--*  t30    int
               [000031] ------------        t31 = *  ADD       int
                                                  /--*  t31    int
               [000032] ------------        t32 = *  PUTARG_TYPE int
               [000033] ------------        t33 =    LCL_VAR   int    V04 tmp3         u:1
               [000034] ------------        t34 =    CNS_INT   int    8
                                                  /--*  t33    int
                                                  +--*  t34    int
               [000035] ------------        t35 = *  ADD       int
                                                  /--*  t35    int
                                                  +--*  t15    ref
               [000036] ------------              *  STOREIND  ref
               [000037] ------------        t37 =    LCL_VAR   int    V04 tmp3         u:1 (last use)
               [000038] ------------        t38 =    CNS_INT   int    12
                                                  /--*  t37    int
                                                  +--*  t38    int
               [000039] ------------        t39 = *  ADD       int
                                                  /--*  t39    int
                                                  +--*  t5     ref
               [000040] ------------              *  STOREIND  ref
                                                  /--*  t32    int    arg-1
                                                  +--*  t9     int    calli tgt
N011 ( 47, 19) [000010] --CXGO------              *  CALL ind nullcheck void
               [000018] ------------                 IL_OFFSET void   IL offset: 0xd
               [000019] ------------                 IL_OFFSET void   IL offset: 0xe
N001 (  0,  0) [000012] ------------                 RETURN    void
```
`t9` is the target, coming from `t9 =    LCL_VAR   int    V02 tmp1         u:1 (last use)` and is the method handle for https://github.com/dotnet/runtimelab/blob/dd8c43a825c41ccfc21a8ceefa79e232adbe02a1/src/libraries/System.Private.CoreLib/src/System/Diagnostics/DebugProvider.cs#L66  

To resolve that, somewhere, and with the IL->LLVM compiler, this would be in the IlToLLVMImporter, this needs to be a look up into the vtable for the runtime type, i.e. `s_provider`/`this`.  Easier to see before LLVM lowering maybe:
```
------------ BB01 [???..???), preds={} succs={BB02}
N001 (  0,  0) [000000] ------------                 NOP       void

------------ BB02 [000..00F) (return), preds={BB01} succs={}
               [000017] ------------                 IL_OFFSET void   IL offset: 0x0
N001 (  1,  4) [000006] H-----------         t6 =    CNS_INT(h) int    0x420058 ftn
                                                  /--*  t6     int
N003 (  5,  7) [000008] DA----------              *  STORE_LCL_VAR int    V02 tmp1
N001 ( 14,  2) [000001] H-CXG-------         t1 =    CALL help byref  HELPER.CORINFO_HELP_READYTORUN_STATIC_BASE
N002 (  1,  1) [000002] ------------         t2 =    CNS_INT   int    4 Fseq[s_provider]
                                                  /--*  t1     byref
                                                  +--*  t2     int
N003 ( 15,  3) [000003] ---XG--N----         t3 = *  ADD       byref
                                                  /--*  t3     byref
N004 ( 17,  5) [000004] V--XGO-N----         t4 = *  IND       ref
                                                  /--*  t4     ref
N006 ( 21,  8) [000014] DA-XGO----L-              *  STORE_LCL_VAR ref    V03 tmp2
N008 (  3,  2) [000015] ------------        t15 =    LCL_VAR   ref    V03 tmp2
N009 (  3,  2) [000005] ------------         t5 =    LCL_VAR   ref    V00 arg0
N010 (  3,  2) [000009] ------------         t9 =    LCL_VAR   int    V02 tmp1
                                                  /--*  t15    ref    this in rcx
                                                  +--*  t5     ref    arg1 in rdx
                                                  +--*  t9     int    calli tgt
N011 ( 47, 19) [000010] --CXGO------              *  CALL ind nullcheck void
               [000018] ------------                 IL_OFFSET void   IL offset: 0xd
N001 (  1,  1) [000011] ------------                 NO_OP     void
               [000019] ------------                 IL_OFFSET void   IL offset: 0xe
N001 (  0,  0) [000012] ------------                 RETURN    void
```
I could attempt that as part of this PR if you like?  Basically find the symbol for the slot, then look that up in the EE type - these LLVM steps
```llvm
 %LoadAddressOfSymbolNode2 = load i32*, i32** @__getslot__S_P_CoreLib_System_Diagnostics_DebugProvider__WriteLine___SYMBOL
%WriteLine_slot = load i32, i32* %LoadAddressOfSymbolNode2
.
.
  %CastPtrthis = bitcast i8* %LoadField_s_provider to void (i8*)***,
  %ldEEType = load void (i8*)**, void (i8*)*** %CastPtrthis, align 4, !dbg !446704
  %__getslot__ = getelementptr void (i8*)*, void (i8*)** %ldEEType, i32 %WriteLine_slot
  %ld__getslot__ = load void (i8*)*, void (i8*)** %__getslot__
 ```
My question for Jan or Michael would be is there something in implementation for other backends that helps us?

Reply:
> I could attempt that as part of this PR if you like?

I think it is best we scope this PR to what it is right now, i. e. to indirect calls. I suspect the case being shown here will "just work" once that has been done.

I looked around a bit, and it appears the method in question is using an implementation of "virtualness" through the `CORINFO_CALL_CODE_POINTER` path, where we use an indirect invoke for a helper which dispatches to the actual target method:
```
// CORINFO_CALL_CODE_POINTER (shared generic code only) :
//   Indicates that the JIT should do an indirect call to the entrypoint given by address, which may be specified
//   as a runtime lookup by CORINFO_CALL_INFO::codePointerLookup.
//   [targetMethodHandle] will not hold a valid value.
//   This flag may be combined with nullInstanceCheck=TRUE for uses of callvirt on methods whose target method can
//   be resolved at compile-time but whose instantiation can be resolved only through runtime lookup.
```
Apparently this path is taken because, at least for my reproduction code, the layout of the VTable is not fixed.

Virtuals implemented via VTables go though [`fgExpandVirtualVtableCallTarget`](https://github.com/dotnet/runtime/blob/5dc1b4c31312b6778b62ee28ba4a2b0c54f34898/src/coreclr/jit/morph.cpp#L9273) which builds the required indirections, so we won't have to do much if anything except build the LLVM for `gtControlExpr` (hopefully).

Reply:
Yes I think I see what you are looking at, `CORINFO_CALL_CODE_POINTER` is the `callInfo->kind`, but I'm investigating that as the comment suggests that's only for generics ("shared generic code only")

Reply:
I wonder if instead of the `LazyVTableSliceProvider` which always returns a `LazilyBuiltVTableSliceNode` and hence never has fixed slots:  https://github.com/dotnet/runtimelab/blob/dd8c43a825c41ccfc21a8ceefa79e232adbe02a1/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/VTableSliceNode.cs#L178-L184

I want something more like the `ScannedVTableProvider` which has https://github.com/dotnet/runtimelab/blob/dd8c43a825c41ccfc21a8ceefa79e232adbe02a1/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs#L271-L294

Reply:
I think this might work a new slice provider and node:
```
    public class LLVMVTableSliceProvider : VTableSliceProvider
    {
        internal override VTableSliceNode GetSlice(TypeDesc type)
        {
            // TODO: move ownership of compiler-generated entities to CompilerTypeSystemContext.
            // https://github.com/dotnet/corert/issues/3873
            if (type.GetTypeDefinition() is Internal.TypeSystem.Ecma.EcmaType)
            {
                return new LLVMVTableSliceNode(type);
            }
            else
                return new LazilyBuiltVTableSliceNode(type);
        }
    }

    class LLVMVTableSliceNode : VTableSliceNode
    {
        public LLVMVTableSliceNode(TypeDesc type) : base(type)
        {
        }

        public override IReadOnlyList<MethodDesc> Slots => new MethodDesc []{};

        public override bool HasFixedSlots => true;
    }
```
to send it the path of `fgExpandVirtualVtableCallTarget`.  Instead of 
https://github.com/dotnet/runtimelab/blob/dd8c43a825c41ccfc21a8ceefa79e232adbe02a1/src/coreclr/jit/morph.cpp#L9539
call something that returns the symbol from a method similar to https://github.com/dotnet/runtimelab/blob/dd8c43a825c41ccfc21a8ceefa79e232adbe02a1/src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/ILToLLVMImporter.cs#L2107-L2113
Then before adding the GT_ADD, add a  GT_CNS_INT (GTF_ICON_CONST_PTR?) and GT_IND 

Is that crazy ?

Reply:
The compiler doesn't build the vtable the way the CoreCLR runtime does (`foreach (method in type) if method.IsVirtual AddToVTable(method)`). Instead, the AOT compiler tracks which virtual methods are actually used and only once the method is actually called virtually, it's added to the vtable. Thanks to this, we compile a lot less methods. But because of this, we don't know slot numbers until we looked at all the callvirts in the program.

You'll get the ScannedVTableProvider if you add `--scan` to the ILC command line (or add `-O` to enable optimizations - optimizations imply running the scanner). It will run a pass over the whole program that will import all the IL and build a whole program view. Once the scanner is done, RyuJIT will be instructed to do vtable lookups and all the offsets get provided through JitInterfaces. But without the scanner running, we can't tell the slots because we haven't seen all the virtual calls yet. So we ask RyuJIT to do the call through a helper instead. Helper is in assembly and we figure out what assembly instructions to put there once everything is compiled.

You can just add `--scan` to unblock yourself now. But the helper is similar to the one that is used for grabbing static bases of types - the implementation strategy could be similar (I assume we would want to emit some LLVM IR for the helper from the ILCompiler.LLVM side). The emission of those helpers can only happen once the main compilation is done; otherwise they'll run into the same problem with slot not being known yet.

Reply:
Look for `ReadyToRunHelperId.VirtualCall` in the ILCompiler source tree (e.g. under `DependencyAnalysis\Target_X64`) to see how the helper looks like.

Reply:
@MichalStrehovsky Thanks, yes I see that running the scanner would be better than my idea, less work and one less indirection.  I think that is the way to unblock this PR in the short term.

I have a problem with the scanner on and
```
GenericDictionaryLookup lookup = _compilation.ComputeGenericLookup(_method, helperId, helperArg);
``` 
returning a lookup with an invalid `HelperId` for the IL->LLVM backend so I'll need to see what's up with that.

Reply:
Hmm mix n match between fixed slots for RyuJit and always helper lookups for the IL backend looks awkward.  I guess I'll need to adapt the IL backend for fixed slots (for GetThreadStaticBase for example).  This code assumes there is always a helper:
```cs
        ISymbolNode GetGenericLookupHelperAndAddReference(ReadyToRunHelperId helperId, object helperArg, out LLVMValueRef helper, IEnumerable<LLVMTypeRef> additionalArgs = null)
        {
            ISymbolNode node;
            GenericDictionaryLookup lookup = _compilation.ComputeGenericLookup(_method, helperId, helperArg);

            var retType = helperId == ReadyToRunHelperId.DelegateCtor
                ? LLVMTypeRef.Void
                : LLVMTypeRef.CreatePointer(LLVMTypeRef.Int8, 0);
            var helperArgs = new List<LLVMTypeRef>
            {
                LLVMTypeRef.CreatePointer(LLVMTypeRef.Int8, 0),
                LLVMTypeRef.CreatePointer(LLVMTypeRef.Int8, 0),
            };
            if (additionalArgs != null) helperArgs.AddRange(additionalArgs);
            if (_method.RequiresInstMethodDescArg())
            {
                node = _compilation.NodeFactory.ReadyToRunHelperFromDictionaryLookup(lookup.HelperId, lookup.HelperObject, _method);
                helper = GetOrCreateLLVMFunction(node.GetMangledName(_compilation.NameMangler),
                    LLVMTypeRef.CreateFunction(retType, helperArgs.ToArray(), false));
            }
            else
            {
                Debug.Assert(_method.RequiresInstMethodTableArg() || _method.AcquiresInstMethodTableFromThis());
                node = _compilation.NodeFactory.ReadyToRunHelperFromTypeLookup(lookup.HelperId, lookup.HelperObject, _method.OwningType);
                helper = GetOrCreateLLVMFunction(node.GetMangledName(_compilation.NameMangler),
                    LLVMTypeRef.CreateFunction(retType, helperArgs.ToArray(), false));
            }
```   
Bit surprised as Release configuration used to work in CoreRT, it would have had the scanner on I would have thought.

Reply:
> returning a lookup with an invalid HelperId for the IL->LLVM backend so I'll need to see what's up with that

Generic lookups have the same problem as vtable lookups - we don't know the slots until after we looked at the whole program - that's why they need to go through helpers. They also have the same solution - scanner. Once the scanner runs, helper calls are no longer needed - the lookup sequence can be inlined - it's similar to how vtable lookup is done - just a couple indirections. The result from `ComputeGenericLookup` tells what offsets from the generic context to indirect. But you can always just ifdef things out and and unblock yourself (e.g. make it always take the helper route).

> Bit surprised as Release configuration used to work in CoreRT, it would have had the scanner on I would have thought.

We used to have the scanner disabled on WASM and CppCodegen: https://github.com/dotnet/corert/blob/c6af4cfc8b625851b91823d9be746c4f7abdc667/src/ILCompiler/src/Program.cs#L652-L658 because it was easier to disable than adding support for the inlined lookups.

Reply:
I can get back to this now.  With the scanner enabled I get different IR
```
------------ BB02 [000..00F) (return), preds={BB01} succs={}
               [000018] ------------                 IL_OFFSET void   IL offset: 0x0
N001 ( 14,  2) [000001] H-CXG-------         t1 =    CALL help byref  HELPER.CORINFO_HELP_READYTORUN_STATIC_BASE
N002 (  1,  1) [000002] ------------         t2 =    CNS_INT   int    4 Fseq[s_provider]
                                                  /--*  t1     byref
                                                  +--*  t2     int
N003 ( 15,  3) [000003] ---XG--N----         t3 = *  ADD       byref
                                                  /--*  t3     byref
N004 ( 17,  5) [000004] V--XGO-N----         t4 = *  IND       ref
                                                  /--*  t4     ref
N006 ( 21,  8) [000010] DA-XGO----L-              *  STORE_LCL_VAR ref    V02 tmp1
N008 (  3,  2) [000011] ------------        t11 =    LCL_VAR   ref    V02 tmp1
N009 (  3,  2) [000005] ------------         t5 =    LCL_VAR   ref    V00 arg0
N010 (  3,  2) [000013] ------------        t13 =    LCL_VAR   ref    V02 tmp1
                                                  /--*  t13    ref
N011 (  6,  4) [000014] #--X--------        t14 = *  IND       int
N012 (  1,  1) [000015] ------------        t15 =    CNS_INT   int    40
                                                  /--*  t14    int
                                                  +--*  t15    int
N013 (  7,  5) [000016] ---X---N----        t16 = *  ADD       int
                                                  /--*  t16    int
N014 (  9,  7) [000017] n--X--------        t17 = *  IND       int
                                                  /--*  t11    ref    this in rcx
                                                  +--*  t5     ref    arg1 in rdx
                                                  +--*  t17    int    control expr
N015 ( 56, 26) [000006] --CXGO------              *  CALLV vt-ind void   System.Diagnostics.DebugProvider.WriteLine
```
I assume this is what we were expecting - instead of the `CALL ind nullcheck void` we now have `CALLV vt-ind ` ?

Reply:
Yes, that looks right to me.

Reply:
>  I'm going to need that aren't I, to get to the method table?

Why? The idea here is that we don't need to do that, the control expression (== function address) has already been computed for us by morph.

Reply:
Ok, cool, this is super simple now it seems, thanks.
![image](https://user-images.githubusercontent.com/2720041/152209385-86c0a00f-7761-4307-9ec2-300b521f95a7.png)
```
        if (call->IsVirtualVtable())
        {
            FunctionType* functionType = createFunctionTypeForCall(call);

            Value* funcPtr = castIfNecessary(getGenTreeValue(call->gtControlExpr), functionType->getPointerTo());

            llvmFuncCallee = {functionType, funcPtr};
        }
```